### DISCUSSION TITLE
Example of Roland ADSR like Envelope graph

### DISCUSSION THREAD

**Discussion by dnaldoog** [RomanKubiak/ctrlr#626] [Show and tell]
An example of an ADSR (type) graph. 
Reacts to Mouse Drag and uiSlider value changes.

![JD-990 TVF Example panel](https://github.com/RomanKubiak/ctrlr/assets/76831683/943eea88-97c7-4fa1-8846-bf8fbf3aa11d)

[Example TVF graph JD990_1.zip](https://github.com/RomanKubiak/ctrlr/files/14737813/Example.TVF.graph.JD990_1.zip)

**Comment by damiensellier** [comment]
That's awesome. Someone on this board is going to be happy... :)

**Comment by dnaldoog** [comment]
Also there is the [standard ADSR &#x2015; on the Juno Alpha panel](https://sourceforge.net/projects/juno-alpha/)

![junoAlphaADSR](https://github.com/RomanKubiak/ctrlr/assets/76831683/6a3057a7-dac3-4570-9333-0e68ce4ddb9e)

**Comment by unityconstruct** [comment]
lol... wasting no time with that snapping sir! 

love that logarithmic envelope line as well... need to check my specs 'cause might be needing such a thing

**Reply by dnaldoog** [reply]
If there was a way to contain that snapping to within the uiComponent, then it could be very useful.

**Reply by Ted-juh** [reply]
Or restrict the snapping of the mouse to a certain area in the panel with the Rectangle and Point class? 

Pseudo-code:
adsrRectangle = Rectangle(x,y,w,h)

Get the coordinates of where the mouse clicks with the Point class.

if Point == adsrRectangle then
Do some mouse snapping because of some cunning code.
End.
 
But this is highly theoretical of course.

**Reply by unityconstruct** [reply]
doh! misread your post... What a great point I hadn't even gotten around to consider [ assuming the prob is that while snapping to node is possible, you then have problem of ONLY snapping to nodes... and no longer can do anything else with the other mods ]

was in this DEMO panel yday looking at ways to data-drive PopupMenu creation ( so one Class/code can serve up dozens of menus instead of requiring `additem` for every option....) and intuit that uiComponent customMouseDown code had a possible key for bounding when snapping would occur. The old HTML days freqently used a bitmap graphic with a `hotzone grid` specifying zones over the bitmap for assigning various hrefs for onClick.. If not assignable per-component, seems a Master Grid might be employed to bound MouseEvent calls... and the component Y/X/WH properties might allow all that to be specified via component reference rather than hard-coded static values
```lua
	if event.mods:isPopupMenu() then  <<---- detection
		ret = m:show(10,5,1,15)

paintCustom = function(comp,g)
	g:fillAll (Colour(0xff0000ff))
	r = Rectangle(0,0,comp:getWidth(),comp:getHeight())  <<----- stealing coord/size from comp props
```
![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/6b76197e-e61c-4a45-b318-bf8f1dc09617)

And was also mulling over some means of using snapping for `movable groups` ... I think the UIX world calls them `drawers` - where you grab a handle and drag to reveal more UIX components.. I like draws but it feels/look ghetto when they don't fully snap in/out automatically when in proximity. 

That JUNO UI looks clean brother!..

**Reply by damiensellier** [reply]
With juce there's some kind of flexbox and component auto snap to grid like in css. You should take a look at those tuto : 

https://docs.juce.com/master/tutorial_flex_box_grid.html
https://docs.juce.com/master/tutorial_rectangle_advanced.html

You're talking about drawers like on phone UI. Do you mean accordions? In Juce they called it "concertina" which makes no sense :) 
https://docs.juce.com/master/classConcertinaPanel.html

EDIT: it makes sense, concertinas are a type of accordion...

**Reply by unityconstruct** [reply]
lol.. concertina wire came to mind... painful like JUCE :)

probably about time i look more at the juce classes before I wade too far off into recreating wheels. Every time I explore, I find a new thing like ModifierKeys & Image()
